pkg/optimizer/datapath_csp.go

```diff
@@ -539,48 +539,49 @@ func (dpc *DataPathCSP) addAnOptimizationGoal(goal adminconfig.AttributeOptimiza
 
 // This creates a param array with the values of the given attribute for each cluster/module/storage account instance
 // NOTE: We currently assume all values are integers. Code should be changed if some values are floats.
-func (dpc *DataPathCSP) getAttributeMapping(attr taxonomy.Attribute) (string, string, error) {
-	instanceType := dpc.env.AttributeManager.GetInstanceType(attr)
-	if instanceType == nil {
+func (dpc *DataPathCSP) getAttributeMapping(attr string) (string, string, error) {
+	instanceTypes := dpc.env.AttributeManager.GetInstanceTypes(attr)
+	if len(instanceTypes) == 0 {
 		return "", "", fmt.Errorf("there are no clusters, modules or storage accounts with an attribute %s", attr)
 	}
-
 	resArray := []string{}
 	varName := ""
-	switch *instanceType {
-	case taxonomy.Cluster:
-		varName = clusterVarname
-		for _, cluster := range dpc.env.Clusters {
-			infraElement := dpc.env.AttributeManager.GetAttribute(attr, cluster.Name)
-			if infraElement == nil {
-				return "", "", fmt.Errorf("attribute %s is not defined for cluster %s", attr, cluster.Name)
+	for _, instanceType := range instanceTypes {
```

Comment:
This is probably not how we want to handle bandwidth-like attributes. We have to check `len(instanceTypes)`. If it's 1, then it's a simple optimization goal and the current code is ok. If it's 2, then it's bandwidth-like optimization goal and we have to `return "", "", nil`

Reply:
This is not related to bandwidth-like attributes. The use-case is a single "cost" attribute defined for both clusters and storage-accounts. When this cost should be minimized it means minimizing everything on the data path that has "cost".
I saw that the code fills an array, so it should not be a problem, or should it?

Reply:
Policies could be about storage cost, cluster cost, or all costs. Minimization should be done on the relevant costs indicated in the policy.  Right?